EOS-26482 : Changes to support hctl status over http command

ydb242 opened the pull request:
Signed-off-by: Yash Bhamare <yash.bhamare@seagate.com>

# Problem Statement
- Currently, cluster status cannot be fetched outside of hare installation. In order to provide this functionality, an endpoint is added in hax to give the cluster status in json format.

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

knekrasov commented on hax/hax/server.py:
```diff
@@ -233,6 +261,7 @@ def _configure(self) -> None:
         app = self._create_server()
         app.add_routes([
             web.get('/', hello_reply),
+            web.get('/hctl_stat', hctl_stat),
```
I think the endpoint should be more reader friendly, e.g. `GET /cluster/status`.
The output should be in JSON rather than a plaintext (because REST endpoints are normally machine-friendly because of possible integrations).

ydb242 replied:
Updated

knekrasov reviewed (commented):
1. We're about to add a new integration point to our REST API. Where are we going to describe it (including the expected structure and format of the output)?
2. Please reword your commit into "problem -> decision" form.
3. Please add some details to your PR (what problem are you going to solve? are there any requirements that you're fulfilling?)

mssawant commented:
@ydb242, please check the rfc folder in Hare sources, there's already some documentation about hctl status (https://github.com/Seagate/cortx-hare/blob/main/rfc/11/README.md), you can update the same or create a new rfc document.

mssawant commented:
@ydb242  Please post a sample test output.

ydb242 commented:
> @ydb242 Please post a sample test output.

@mssawant I am trying this using the kubernetes branches of motr+hare cluster. As a result, m0d processes dont work but hctl status gives us output for required testing.
Following is the output:

> [root@ssc-vm-g2-rhev4-0108 ~]# curl http://192.168.55.173:8008/cluster/status
{
  "pools": [
    {
      "fid": "0x6f00000000000001:0x2f",
      "name": "the pool"
    }
  ],
  "profiles": [
    {
      "fid": "0x7000000000000001:0x4f",
      "name": "default",
      "pools": [
        "the pool",
        null,
        null
      ]
    }
  ],
  "filesystem": {
    "stats": {}
  },
  "nodes": [
    {
      "name": "localhost",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x6",
          "ep": "inet:tcp:192.168.55.173@2001",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x9",
          "ep": "inet:tcp:192.168.55.173@3001",
          "status": "unknown"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0xc",
          "ep": "inet:tcp:192.168.55.173@3002",
          "status": "unknown"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x29",
          "ep": "inet:tcp:192.168.55.173@5001",
          "status": "unknown"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x2c",
          "ep": "inet:tcp:192.168.55.173@5002",
          "status": "unknown"
        }
      ]
    }
  ]
}[root@ssc-vm-g2-rhev4-0108 ~]#

ydb242 commented:
> @ydb242 Please post a sample test output.

Tested the given changes with kubernetes deployment. It is working as expected.

> [root@ssc-vm-g2-rhev4-2494 k8_cortx_cloud]# curl http://172.16.140.156:8008/cluster/status
{
  "pools": [
    {
      "fid": "0x6f00000000000001:0x81",
      "name": "storage-set-1__sns"
    }
  ],
  "profiles": [
    {
      "fid": "0x7000000000000001:0xcc",
      "name": "Profile_the_pool",
      "pools": [
        "storage-set-1__sns",
        "storage-set-1__dix",
        null
      ]
    }
  ],
  "filesystem": {
    "stats": {
      "fs_free_seg": 429501495192,
      "fs_total_seg": 429502830880,
      "fs_free_disk": 858993459200,
      "fs_avail_disk": 858993459200,
      "fs_total_disk": 858993459200,
      "fs_svc_total": 12,
      "fs_svc_replied": 12
    },
    "timestamp": 1639393339.165318,
    "date": "2021-12-13T11:02:19.165318"
  },
  "nodes": [
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-1910",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x31",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@2001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x34",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x41",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x4e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x51",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@4001",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x54",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@4002",
          "status": "started"
        }
	  ]
    },
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-2493",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x7",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@2001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0xa",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x17",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x24",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x27",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@4001",
          "status": "started"
        },
		{
          "name": "s3server",
          "fid": "0x7200000000000001:0x2a",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@4002",
          "status": "started"
        }
      ]
    },
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-2494",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x5b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@2001",
          "status": "started"
        },
		{
          "name": "ioservice",
          "fid": "0x7200000000000001:0x5e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x6b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x78",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x7b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@4001",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x7e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@4002",
          "status": "started"
        }
      ]
    }
  ]
}[root@ssc-vm-g2-rhev4-2494 k8_cortx_cloud]#

ydb242 commented:
> @ydb242, please check the rfc folder in Hare sources, there's already some documentation about hctl status (https://github.com/Seagate/cortx-hare/blob/main/rfc/11/README.md), you can update the same or create a new rfc document.

Updated the 11/HCTL RFC to include information about the new endpoint.

knekrasov reviewed (commented):
1. Can you squash your commits into one please?
2. Please reword the PR name. "Initial commit" doesn't look well anyway.
3. Can you also post the output of curl with headers (i.e. `curl -i http://<youserver>:8008/cluster/status`)? I just wonder if the generated output has the application/json Content-type.

knekrasov commented:
@ydb242 
> I am planning to use squash and merge option of the github PR. It will make sure that all the commits are squashed into one while merging.

We normally don't do that in Hare component (so we keep having separate commits in the target branch so they can be cherry-picked back and forth - you know how many branches we may need to maintain 😉 ).
So the idea is that you as the PR author is the best person to decide which commit deserves to exist as a separate one in the main historical line and which ones are a kind of noise (hesitations, debug stuff, scaffolding elements etc) and thus should be melded into the "real" commits.

cc @mssawant, @suvratjoshi

ydb242 commented:
> @ydb242
> 
> > I am planning to use squash and merge option of the github PR. It will make sure that all the commits are squashed into one while merging.
> 
> We normally don't do that in Hare component (so we keep having separate commits in the target branch so they can be cherry-picked back and forth - you know how many branches we may need to maintain 😉 ). So the idea is that you as the PR author is the best person to decide which commit deserves to exist as a separate one in the main historical line and which ones are a kind of noise (hesitations, debug stuff, scaffolding elements etc) and thus should be melded into the "real" commits.
> 
> cc @mssawant, @suvratjoshi

Ok. I get the point.
I have updated the branch, and squashed the multiple commits into one.

mssawant commented:
@ydb242, looks good. Can you also try with hax pod service name instead of an explicit ip address? I mean it should be possible to fetch status from any given cortx data container hostname/service name as hax run in all the data pods.
e.g. curl http://cortx-data-headless-svc-ssc-vm-g2-rhev4-1910:8008/cluster/status

mssawant commented:
@ydb242 , commit message does not look right, consider,
```
Problem: no way to fetch hctl status outside of Hare installation

Hare implements a `hctl` control interface command to fetch Hare and Motr cluster status. This interface requires cortx-hare software to be installed which adds to user dependencies.

Solution:
- Start a http server in Hare hax service
- Implement a http command to fetch Hare and Motr cluster status.
```

ydb242 commented:
> @ydb242 , commit message does not look right, consider,
> 
> ```
> Problem: no way to fetch hctl status outside of Hare installation
> 
> Hare implements a `hctl` control interface command to fetch Hare and Motr cluster status. This interface requires cortx-hare software to be installed which adds to user dependencies.
> 
> Solution:
> - Start a http server in Hare hax service
> - Implement a http command to fetch Hare and Motr cluster status.
> ```

Updated

ydb242 commented:
> @ydb242, looks good. Can you also try with hax pod service name instead of an explicit ip address? I mean it should be possible to fetch status from any given cortx data container hostname/service name as hax run in all the data pods. e.g. curl http://cortx-data-headless-svc-ssc-vm-g2-rhev4-1910:8008/cluster/status

Yes, I tried the above pattern of hostname/service with all the containers in the data pod. I was able to retrieve the cluster status.
I even logged into the control pod and accessed the csm-agent container. From there too, I was able to retrieve cluster status using hostname/service pattern.

> [root@cortx-control-headless-svc /]# hostname
cortx-control-headless-svc
[root@cortx-control-headless-svc /]# curl http://cortx-data-headless-svc-ssc-vm-g2-rhev4-2494:8008/cluster/status
{
  "pools": [
    {
      "fid": "0x6f00000000000001:0x81",
      "name": "storage-set-1__sns"
    }
  ],
  "profiles": [
    {
      "fid": "0x7000000000000001:0xcc",
      "name": "Profile_the_pool",
      "pools": [
        "storage-set-1__sns",
        "storage-set-1__dix",
        null
      ]
    }
  ],
  "filesystem": {
    "stats": {
      "fs_free_seg": 429501495192,
      "fs_total_seg": 429502830880,
      "fs_free_disk": 858993459200,
      "fs_avail_disk": 858993459200,
      "fs_total_disk": 858993459200,
      "fs_svc_total": 12,
      "fs_svc_replied": 12
    },
    "timestamp": 1639463271.040964,
    "date": "2021-12-14T06:27:51.040964"
  },
  "nodes": [
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-1910",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x31",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@2001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x34",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x41",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x4e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x51",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@4001",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x54",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1910@4002",
          "status": "started"
        }
      ]
    },
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-2493",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x7",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@2001",
          "status": "started"
        },
		{
          "name": "ioservice",
          "fid": "0x7200000000000001:0xa",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x17",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x24",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x27",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@4001",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x2a",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2493@4002",
          "status": "started"
        }
      ]
    },
    {
      "name": "cortx-data-headless-svc-ssc-vm-g2-rhev4-2494",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x5b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@2001",
          "status": "started"
		},
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x5e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3001",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0x6b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3002",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x78",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@3003",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x7b",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@4001",
          "status": "started"
        },
        {
          "name": "s3server",
          "fid": "0x7200000000000001:0x7e",
          "ep": "inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2494@4002",
          "status": "started"
        }
      ]
    }
  ]
}

mssawant reviewed (approved):
Great work @ydb242!
